File: app/src/main/java/com/example/androidudpserver/MainActivity.java
package com.example.androidudpserver;

import android.os.Bundle;
import android.util.Log;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.Date;
import java.util.Enumeration;
import java.net.ServerSocket;
import java.net.Socket;
import java.io.*;


public class MainActivity extends AppCompatActivity {

    private final static String TAG = "asd"; //MainActivity.class.getSimpleName();

    TextView infoIp, infoPort;
    TextView textViewState, textViewPrompt;

    static final int tcpPORT = 52222;
    TCPServerThread tcpServer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

    }

    @Override
    protected void onStart() {
        tcpServer = new TCPServerThread();
        tcpServer.start();
        super.onStart();
    }

    @Override
    protected void onStop()
    {
        if(tcpServer != null)
        {
            tcpServer = null;
        }

        super.onStop();
    }

    private class TCPServerThread extends Thread
    {
        ServerSocket server;
        ClientWorker w;

        public TCPServerThread() {
            super();
            listenSocket();
        }

        public void listenSocket() {
            try {
                server = new ServerSocket(tcpPORT);
                server.setReuseAddress(true);
            } catch (IOException e) {
                System.out.println("Could not listen on port 4444");
                System.exit(-1);
            }

            while (true) {
                try {
                    w = new ClientWorker(server.accept());
                    Thread t = new Thread(w);
                    t.start();
                } catch (IOException e) {
                    System.out.println("Accept failed: 52222");

                    System.exit(-1);
                }
            }
        }


        class ClientWorker implements Runnable {
            private Socket client;

            /*Constructor*/
            ClientWorker(Socket client) {
                this.client = client;
            }

            @Override
            public void run() {
                String line;
                BufferedReader in = null;
                PrintWriter out = null;

                try {
                    in = new BufferedReader(new InputStreamReader(client.getInputStream()));
                    out = new PrintWriter(client.getOutputStream(), true);
                } catch (IOException e) {
                    System.out.println("in or out failed");
                    System.exit(-1);
                }

                while (true) {
                    try {
                        line = in.readLine();

                        out.println(line);
                    } catch (IOException e) {
                        System.out.println("Read failed");
                        System.exit(-1);
                    }
                }
            }
        }

    }
}

File: app/src/main/java/com/example/androidudpserver/MainActivity.kt
package com.example.androidudpserver

import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.ServerSocket
import java.net.Socket

class MainActivity : AppCompatActivity(),DevicesAdapter.DeviceItemClickListener {

    var tcpServer: TCPServerThread? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        Log.d("TAG", "starting act")
        val devicesAdapter = DevicesAdapter(this)
        val recView = findViewById<RecyclerView>(R.id.rvDevices)
        recView.layoutManager = LinearLayoutManager(this)
        recView.adapter = devicesAdapter

    }

    override fun onStart() {
        super.onStart()

        val thread = Thread(Runnable {
            try {
                tcpServer = TCPServerThread()
                tcpServer!!.start()
                Log.e("TAG", "starting thread")
                Toast.makeText(applicationContext," Starting Server from Main", Toast.LENGTH_LONG).show()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }).start()


    }

    override fun onStop() {
        super.onStop()
        if (tcpServer != null) {
            tcpServer = null
        }
    }

    inner class TCPServerThread : Thread() {
        private var server: ServerSocket? = null
        private var w: ClientWorker? = null
        private var count:Int = 0
        private val message = "lol"

        init{
            listenSocket()
        }

        private fun listenSocket() {
            try
            {
                server = ServerSocket(tcpPORT)
                server!!.reuseAddress = true
            }
            catch (e: IOException)
            {
                Log.e("TAG", "Cannot listen on the specified port!")
                System.exit(-1)
            }

            while (true)
            {
                try
                {
                    w = ClientWorker(server!!.accept())
                    Log.e("TAG", "Client Accept done")
                    val t = Thread(w)
                    t.start()
                }
                catch (e: IOException)
                {
                    println("Accept failed: 52222")
                    System.exit(-1)
                }
            }
        }
    }
        inner class ClientWorker /*Constructor*/(private val client: Socket) : Runnable {
            override fun run() {
                var line: String
                var input: BufferedReader? = null
                var out: PrintWriter? = null
                try {
                    input = BufferedReader(InputStreamReader(client.getInputStream()))
                    Log.e(TAG, input.toString())
                    out = PrintWriter(client.getOutputStream(), true)
                } catch (e: IOException) {
                    println("in or out failed")
                    System.exit(-1)
                }

                while (true) {
                    try {
                        line = input!!.readLine()
                        Log.e(TAG, line.toString())
                        out!!.println(line)
                    } catch (e: IOException) {
                        println("Read failed")
                        System.exit(-1)
                    }
                }
            }
        }

    companion object {
        private const val TAG = "asd" //MainActivity.class.getSimpleName();
        const val tcpPORT = 4446
    }

    override fun onItemClick(clickedStore: Device) {
        TODO("Not yet implemented")

    }
}